### ISSUE TITLE
[Feature] Tensor Parallel (TP)

### ISSUE THREAD

**Issue by namgyu-youn** [facebookresearch/optimizers#157]
I am excited to deepen my contributions after slightly understanding the Kronecker-product preconditioner. As you know, multiple distribution strategies for the shampoo focus on data parallelism and parameter sharding.

And Tensor Parallelism (TP) may offer substantial benefits in two specific contexts:
- Optimization of Kronecker-product updates
- Load balancing during matrix inversion

Before proceeding, I would like to ask whether the TP aligns with the core objectives of the research. I hypothesize that, by introducing TP, we could unlock a broader spectrum of advanced workflows—such as combining DDP with TP.

**Comment by namgyu-youn** [comment]
> I am not familiar with Tensor Parallel (TP) but if we could further speedup this. It will be cool. The last time we ran performance benchmark on the current implementation is bounded on [the `torch.tensordot` operations](https://github.com/facebookresearch/optimizers/blob/main/distributed_shampoo/utils/shampoo_preconditioner_list.py#L884-L899) which we could not fused together ([@gajjanag](https://github.com/gajjanag) , correct me if I were wrong on this).
> 
> One thing [@runame](https://github.com/runame) and I talked few months ago is trying `torch.einsum` for the `torch.tensordot` operation to see if `torch.einsum` could benefit from that. However, [@runame](https://github.com/runame) also told me we should have benchmarks ready before doing experiments like this, which I agree on that.
> 
> How about we start from building benchmarks first, and that will be a good way to gauge on how all those ideas, e.g., TP, `torch.einsum`, etc, work?

Thanks for your full background explanation! I have thought there should be a benchmark in the internal system because this is the research project, but it seems there isn't yet. For a deeper understanding, I would like to ask some questions:

1. Is there any chance to get some credit for multi-GPUs? Since I am using a rental service at RunPod (2 X V100, 1 X L40s), the cost issue might happen while building a benchmark. I am ready to dive into it, but cost issues are the challenge.
2. For deeper understanding, could you share some internal documents related to the benchmark (e.g., base model, evaluation metrics, multiple-GPU environments)? Since building a benchmark requires full understanding, it might be a useful resource for me.

**Comment by namgyu-youn** [comment]
> 1. Unfortunately there is no credit for multi-GPUs but it is also good news for the problem I mentioned, it is single node only problem. Basically all the computations done at `shampoo_preconditioner_list.py` are happened on a single node only.

Then I am deeply interested in single-GPU benchmarks.

> 2. Internal benchmarks are probably based on internal models so it could not be shared. However, I do believe what we have now is totally transferrable improvements if we could make any. What I am looking for is a general improvement rather than specializations on certain models. It is similar to [Update factor_metrics logic for `BaseShampooPreconditionerList` #144](https://github.com/facebookresearch/optimizers/pull/144) and [Consolidate multiple NaN/Inf check into single check #117](https://github.com/facebookresearch/optimizers/pull/117) you did before but bigger.

In my thought, a benchmark would contain the following elements, focusing on comparison (e.g., FSDP vs HSDP) :
- Performance estimation (e.g., MACs, FLOPs, GPU-usage): Personally, [torchprofile](https://github.com/zhijian-liu/torchprofile) was useful for evaluating MACs and FLOPs.
- Dataset (e.g., ImageNet, CIFAR-10) and base model (i.e., already **public** architecture, not internal (private) model)

And it seems the benchmark in this project is not the same as above. Could you elaborate more about it? #117 and #144 were intended for general improvement, but they are based on theory, not on a chart or table with an experiment.

**Comment by namgyu-youn** [comment]
> > In my thought, a benchmark would contain the following elements, focusing on comparison (e.g., FSDP vs HSDP) :
> > 
> > * Performance estimation (e.g., MACs, FLOPs, GPU-usage): Personally, [torchprofile](https://github.com/zhijian-liu/torchprofile) was useful for evaluating MACs and FLOPs.
> > * Dataset (e.g., ImageNet, CIFAR-10) and base model (i.e., already **public** architecture, not internal (private) model)
> > 
> > And based on your comment, it seems this is not objective. Could you elaborate more about it? [#117](https://github.com/facebookresearch/optimizers/pull/117) and [#144](https://github.com/facebookresearch/optimizers/pull/144) were intended for general improvement, but they are based on theory, not real experiment with charts or tables. Or are you looking to use dummy Tensor?
> 
> For comparisons of FSDP vs. HSDP, that definitely requires more setup. I think a good small scale benchmark focusing on single node computation will be like [8f6c523](https://github.com/facebookresearch/optimizers/commit/8f6c5232c2e83b8c1c9b5fad8e49219d2e22907c) and [8908bf2](https://github.com/facebookresearch/optimizers/commit/8908bf294e2c193a567d5ff5f5c1de87305eb17a). For [#117](https://github.com/facebookresearch/optimizers/pull/117) and [#144](https://github.com/facebookresearch/optimizers/pull/144), we definitely could do something like the examples I showed before but personally I am pretty convinced yours are just no worse than what we had before, and moreover, yours are simply better for code tidiness perspective.
> 
> I think a concrete potential goal personally I want to know is that does `torch.einsum()` perform better than `torch.tensordot()` on https://github.com/facebookresearch/optimizers/blob/main/distributed_shampoo/utils/shampoo_preconditioner_list.py#L884-L899. If we could find a way to rewrite that part based on `torch.einsum()`, we could build a small benchmark like [8908bf2](https://github.com/facebookresearch/optimizers/commit/8908bf294e2c193a567d5ff5f5c1de87305eb17a) and verify that.
> 
> Hope this give you a concrete example, and this is the reason I am saying although this is a general improvement without any model required.

Got it. Now I see the reason there is no need for multi-GPUs and specific modeling. I really appreciate your full explanation about affordable benchmarks I can try. Before the start, I want to check the following list. Or is there something I should care about? 
- [torchprofile](https://github.com/zhijian-liu/torchprofile): This was useful for calculating MACs or FLOPs.
- GPU usage: `nvidia-smi` or `torch.cuda.memory_allocated()` could be a choice.
- Visualization: Chart visualization would be useful for understanding full process.

**Comment by namgyu-youn** [comment]
> Hope this give you a concrete example, and this is the reason I am saying although this is a general improvement without any model required.

Could you review [4fc5769](https://github.com/namgyu-youn/optimizers/commit/4fc576975a3b8bb012a8b0d7e7fd0d530f3f5d5e)? This is the primitive benchmark for the preconditioner, and the folder structure is placed in the root directory because it is only needed for the developer side.

**Comment by namgyu-youn** [comment]
> > Thanks to your full explanation, now I believe I can handle this task. Also, now I understood the reason why there is no need for multi-GPUs and specific modeling. Before the start, could you comment on the following list? Let me know if there is anything I should care about or if I am missing something.
> > 
> > * [torchprofile](https://github.com/zhijian-liu/torchprofile): This was useful for calculating MACs or FLOPs.
> > * GPU usage: `nvidia-smi` or `torch.cuda.memory_allocated()` could be a choice.
> > * Visualization: Chart visualization would be useful for understanding the full process. (save into CSV could be a choice.)
> > * Axis : `ShampooPreconditionerList`, `EigendecomposedShampooPreconditionerList`, `AdagradPreconditionerList`
> > 
> > Also, I want to ask if benchmarks for operations in `matrix_functions` (e.g., `matrix_eigendecomposition`, `matrix_inverse_root`) could be useful because they are core computations for preconditioners.
> 
> Yes, those will be good for sure. Right now I will not worry about visualization so much because normal text-based output are fine.

Got it. Since log messages could be a choice for text-based output, is there no need to save into CSV? Then it could be more simpler.
